### PR TITLE
release 0.2.0

### PR DESCRIPTION
**Pull request by chdsbd**
CHANGES
- throw ValueError if carriage return is in input
- combine adjacent spans
- bump version
- add dockerfile
- setup circleci to lint and test project
- document build and release process

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -12,22 +13,56 @@ fn markdown_html_finder(_py: Python, m: &PyModule) -> PyResult<()> {
     /// Find positions of html nodes in a markdown file
     ///
     /// Returns a list of tuples of start and end positions
-    fn find_html_positions_py(markdown: &str) -> PyResult<Vec<(usize, usize)>> {
-        Ok(find_html_positions(markdown))
+    /// 
+    /// Raises ValueError when passed carriage returns. You must strip carriage
+    /// returns before calling this function.
+    fn find_html_positions_py(markdown: &'static str) -> PyResult<Vec<(usize, usize)>> {
+        match find_html_positions(markdown) {
```
Comment (sbdchd) [inline src/lib.rs]:
```suggestion
       find_html_positions(markdown).map_err(exceptions::ValueError::py_err)
```

File: src/lib.rs [hunk 2]
```diff
@@ -12,22 +13,56 @@ fn markdown_html_finder(_py: Python, m: &PyModule) -> PyResult<()> {
...
     }
 
     Ok(())
 }
 
-fn find_html_positions(markdown: &str) -> Vec<(usize, usize)> {
-    Parser::new_ext(markdown, Options::empty())
+fn join_adjacent_spans(spans: Vec<(usize, usize)>) -> Vec<(usize, usize)> {
+    let mut new_spans = Vec::new();
+    for span in spans.iter() {
```
Comment (sbdchd) [inline src/lib.rs]:
```suggestion
    for (span_start, span_end) in spans.iter() {
```

File: src/lib.rs [hunk 3]
```diff
@@ -12,22 +13,53 @@ fn markdown_html_finder(_py: Python, m: &PyModule) -> PyResult<()> {
...
+        let (span_start, span_end) = span;
+        if let Some(prev_span) = new_spans.last_mut() {
+            let (prev_start, prev_end) = prev_span;
+            if prev_end == span_start {
```
Comment (sbdchd) [inline src/lib.rs]:
```suggestion
            if prev_end >= span_start {
```

Comment (chdsbd) [reply src/lib.rs]: This would never happen in this case, but I'll add a test anyway.

File: src/lib.rs [hunk 4]
```diff
@@ -12,22 +13,52 @@ fn markdown_html_finder(_py: Python, m: &PyModule) -> PyResult<()> {
     /// Find positions of html nodes in a markdown file
     ///
     /// Returns a list of tuples of start and end positions
-    fn find_html_positions_py(markdown: &str) -> PyResult<Vec<(usize, usize)>> {
-        Ok(find_html_positions(markdown))
+    ///
+    /// Raises ValueError when passed carriage returns. You must strip carriage
+    /// returns before calling this function.
+    fn find_html_positions_py(markdown: &'static str) -> PyResult<Vec<(usize, usize)>> {
+        find_html_positions(markdown).map_err(|msg| exceptions::ValueError::py_err(msg))
```
Comment (sbdchd) [inline src/lib.rs]:
```suggestion
        find_html_positions(markdown).map_err(exceptions::ValueError::py_err)
```

**Review by sbdchd — APPROVED** [review]

**Pull request merged**